src/endless/Eosldr.cpp

```diff
@@ -0,0 +1,204 @@
+#include "stdafx.h"
+#include "Eosldr.h"
+#include "GeneralCode.h"
+#include "WMI.h"
+
+// Paths for g2ldr binaries, used to chainload GRUB from the Windows Boot
+// Manager on BIOS systems, rather than the other way round.
+#define EOSLDR_SUBDIRECTORY             L"eosldr"
+// Stub loaded by WBM
+#define EOSLDR_MBR_FILE                 L"eosldr.mbr"
+// GRUB core.img with a special preamble, loaded by eosldr.mbr
+#define EOSLDR_IMG_FILE                 L"eosldr"
+
+// GUID for Endless OS's entry in the BCD boot menu, stored in the Uninstall
+// registry key
+#define REGKEY_BCD_GUID         L"EndlessBcdGuid"
+
+bool EosldrInstaller::CanInstall(const CString & bootFilesPath)
+{
+    FUNCTION_ENTER_FMT("%ls", bootFilesPath);
+
+    return !!PathFileExists(bootFilesPath + EOSLDR_SUBDIRECTORY);
+}
+
+static bool CopyAndHide(const CString &source, const CString &target)
+{
+    FUNCTION_ENTER_FMT("%ls -> %ls", source, target);
+
+    IFFALSE_RETURN_VALUE(
+        CopyFileEx(source, target, NULL, NULL, NULL, 0),
+        "CopyFileEx failed", false);
+    IFFALSE_RETURN_VALUE(
+        SetFileAttributes(target, 
+            FILE_ATTRIBUTE_READONLY |
+            FILE_ATTRIBUTE_HIDDEN |
+            FILE_ATTRIBUTE_SYSTEM),
+        "SetFileAttributes failed", false);
+
+    return true;
+}
+
+bool EosldrInstaller::Install(const CString & systemDriveLetter, const CString & bootFilesPath)
+{
+    const CString eosldrImgPath = systemDriveLetter + EOSLDR_IMG_FILE;
+    const CString eosldrMbrPath = systemDriveLetter + EOSLDR_MBR_FILE;
+
+    const CString eosldrSourceDir = bootFilesPath + EOSLDR_SUBDIRECTORY L"\\";
+    const CString eosldrImgSource = eosldrSourceDir + EOSLDR_IMG_FILE;
+    const CString eosldrMbrSource = eosldrSourceDir + EOSLDR_MBR_FILE;
+
+    if (!(
+        CopyAndHide(eosldrImgSource, eosldrImgPath) &&
+        CopyAndHide(eosldrMbrSource, eosldrMbrPath)
+        )) {
+        return false;
+    }
+
+    return AddToBootOrder(systemDriveLetter, eosldrMbrPath);
+}
+
+bool EosldrInstaller::IsInstalled(const CString & systemDriveLetter)
+{
+    FUNCTION_ENTER_FMT("%ls", systemDriveLetter);
+
+    return !!PathFileExists(systemDriveLetter + EOSLDR_MBR_FILE);
+}
+
+static bool UnhideAndDelete(const CString &path)
+{
+    FUNCTION_ENTER_FMT("%ls", path);
+
+    IFFALSE_PRINTERROR(SetFileAttributes(path, FILE_ATTRIBUTE_NORMAL),
+        "SetFileAttributes failed; will try to delete anyway", false);
+    IFFALSE_RETURN_VALUE(DeleteFile(path), "DeleteFile failed", false);
+
+    return true;
+}
+
+bool EosldrInstaller::Uninstall(const CString & systemDriveLetter)
+{
+    const CString eosldrImgPath = systemDriveLetter + EOSLDR_IMG_FILE;
+    const CString eosldrMbrPath = systemDriveLetter + EOSLDR_MBR_FILE;
+
+    // TODO: just warn on failure?
```

Comment:
I think it's right to make this function consistent with IsInstalled - so either we should check the boot order in that function, or be happy to not removing the boot entry in this case. I'm leaning towards the latter seeing as leaking a boot entry probably isn't the worst thing in the world, compared to complicating the IsInstalled path and incorrectly entering the MBR-mangling code? Or maybe we want IsInstalled to be true if either the file or the boot order are present, and neither to fail the uninstall?